src/libdredd/src/mutate_visitor.cc

```diff
@@ -55,17 +55,29 @@ bool MutateVisitor::TraverseDecl(clang::Decl* decl) {
     // This is the top-level translation unit declaration, so descend into it.
     return RecursiveASTVisitor::TraverseDecl(decl);
   }
-  if (GetSourceRangeInMainFile(compiler_instance_.getPreprocessor(), *decl)
-          .isInvalid()) {
+  auto source_range_in_main_file =
+      GetSourceRangeInMainFile(compiler_instance_.getPreprocessor(), *decl);
+  if (source_range_in_main_file.isInvalid()) {
     // This declaration is not wholly contained in the main file, so do not
     // consider it for mutation.
     return true;
   }
-  if (first_decl_in_source_file_ == nullptr) {
+  clang::BeforeThanCompare<clang::SourceLocation> comparator(
+      compiler_instance_.getSourceManager());
+  if (start_location_of_first_decl_in_source_file_.isInvalid() ||
+      comparator(source_range_in_main_file.getBegin(),
+                 start_location_of_first_decl_in_source_file_)) {
     // This is the first declaration wholly contained in the main file that has
-    // been encountered: record it so that the dredd prelude can be inserted
-    // before it.
-    first_decl_in_source_file_ = decl;
+    // been encountered so far: record it so that the dredd prelude can be
+    // inserted before it.
+    //
+    // The order in which declarations appear in the source file may not exactly
+    // match the order they are visited in the AST (for example, a typedef
+    // declaration is visited after the associated type declaration, even though
+    // the 'typedef' keyword occurs first in the AST), thus this location is
+    // updated each a declaration that appears earlier is encountered.
```

Comment:
Should probably say "each time a declaration that appears earlier is encountered."

Reply:
Thanks!